client/state/query-client.ts

```diff
@@ -4,6 +4,30 @@ import { persistQueryClient } from 'react-query/persistQueryClient-experimental'
 import { shouldPersist, MAX_AGE, SERIALIZE_THROTTLE } from 'calypso/state/initial-state';
 import { getPersistedStateItem, storePersistedStateItem } from 'calypso/state/persisted-state';
 
+const SHOULD_PERSIST_QUERY_KEY = 'shouldPersistQuery';
+
+type PersistencePredicate< T > = ( data: T ) => boolean;
+
+type PersistenceMetaPartial< T > = {
+	[ SHOULD_PERSIST_QUERY_KEY ]: PersistencePredicate< T > | boolean;
+};
+
+const hasPersistenceSetting = ( value: unknown ): value is boolean => {
+	return typeof value === 'boolean';
+};
+
+const hasPersistenceCustomSetting = (
+	value: unknown
+): value is PersistencePredicate< unknown > => {
+	return typeof value === 'function';
+};
+
+export const shouldPersistQuery = < T >(
```

Comment:
Would love to get your thoughts on preferring to invoke a helper function better than setting a custom property.

In other words, why this:

```
useQuery( queryKey, fetchFn, {
	meta: shouldPersistQuery( false ),
} );
```

Is better than:

```
useQuery( queryKey, fetchFn, {
	meta: {
		persist: false,
	},
} );
```

considering that the object signature could essentially allow us to more easily add other options in the future, like for example whether to persist unsuccessful results or specify any other persistence options (like a different lifetime or a more specific cache invalidation strategy for example).

What are your thoughts?

Reply:
The rationale behind a helper function is to hide the property itself. If we ever want to change the property itself, we only need to change one place. I do agree that we're pretty limited in terms of options.

One possibility would be to, in the future, allow a boolean, a function (so we can decide based on the data), and an object with more properties inside.

The idea with a simple boolean is to, as explained in the description, allow for a more general behavior where we don't consider the data but simply want to discard the response, regardless of the content or status.

If we want to add more options in the future, I believe passing an object as the argument would be a good solution.

We could move from a function to a simple constant that's referenced as the property key, I'm open to both. I think that having a function allows for encapsulation of the specified behavior (e.g. you don't have to pass multiple properties to `meta`. Only the factory function is enough and will provide a strong contract on limited constraints).

Reply:
Thanks for elaborating. 

To be honest, I don't think we achieve any positive purpose by hiding the property; instead, we needlessly encapsulate the underlying logic. If we were relying on encapsulation as a method for concealing internal module functionality, it would be a different deal, but that's not the case here. So to me, that's a case where I'd prefer the simple object as an argument.

In terms of providing multiple vs a single property to `meta`, I think we can still continue providing a single property, but let that be more flexible - allow it to be a boolean, an object, or a function. How does that sound?

It would be great if others could chime in here too, this is a discussion after all 😉